src/app/api/upload/route.ts

```diff
@@ -0,0 +1,98 @@
+import { md5 } from "@/lib/md5";
+import { apiError, concat } from "@/lib/utils";
+import withUser from "@/lib/with-user";
+import { NextResponse } from "next/server";
+import { ok, err } from "@justmiracle/result";
+import { R2 } from "@/lib/r2";
+import { db } from "@/db";
+import { user as userTable, user_file } from "@/db/schema";
+import { eq } from "drizzle-orm";
+import { generateId } from "lucia";
+
+// 500MB
+const STORAGE_LIMIT = 500 * 1024 * 1024;
+
+export const POST = withUser(async ({ req, user }) => {
+  const formData = await req.formData().then(ok).catch(err);
+
+  if (formData.error) {
+    return apiError({
+      message: "Something went wrong, please try again later",
+      detailedMessage: formData.error.message,
+      status: 500,
+    });
+  }
+
+  const file = formData.value.get("file");
+
+  if (!(file instanceof File)) {
+    return apiError({
+      message: "Invalid file",
+      status: 400,
+    });
+  }
+
+  const newStorageUsage = user.storageUsage + file.size;
+
+  if (newStorageUsage > STORAGE_LIMIT) {
+    return apiError({
+      message: "Storage limit exceeded",
+      status: 400,
+    });
+  }
+
+  // e.g. image/png -> png
+  const fileExtenstion = file.type.split("/")[1] || "";
+
+  // we can do compression or resizing here and return a new buffer
+  const buffer = Buffer.from(await file.arrayBuffer());
+
+  // hash the content and append the file extension
+  const hashedFilename = concat(md5(buffer), ".", fileExtenstion);
```

Comment:
Change MD5 to SHA1. The change because if we run on the edge, we cannot use NodeJS crypto. We can only use WebCrypto. WebCrypto does not have md5. https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

Reply:
Done